ui/components/app/import-token/token-search/token-search.component.js

```diff
@@ -1,87 +1,87 @@
+  tokenList,
+  searchClassName,
+}) {
+  const t = useI18nContext();
 
-  static propTypes = {
-    onSearch: PropTypes.func,
-    error: PropTypes.string,
-    tokenList: PropTypes.object,
-    searchClassName: PropTypes.string,
-  };
+  const [searchQuery, setSearchQuery] = useState('');
 
-  state = {
-    searchQuery: '',
-  };
+  const [tokenSearchFuse, setTokenSearchFuse] = useState(
+    createTokenSearchFuse(tokenList),
+  );
 
-  constructor(props) {
-    super(props);
-    const { tokenList } = this.props;
-    this.tokenList = Object.values(tokenList);
-    this.tokenSearchFuse = new Fuse(this.tokenList, {
-      shouldSort: true,
-      threshold: 0.45,
-      location: 0,
-      distance: 100,
-      maxPatternLength: 32,
-      minMatchCharLength: 1,
-      keys: [
-        { name: 'name', weight: 0.5 },
-        { name: 'symbol', weight: 0.5 },
-      ],
-    });
-  }
+  useEffect(() => {
+    setTokenSearchFuse(createTokenSearchFuse(tokenList));
+  }, [tokenList, searchQuery]);
 
-  handleSearch(searchQuery) {
-    this.setState({ searchQuery });
-    const fuseSearchResult = this.tokenSearchFuse.search(searchQuery);
-    const addressSearchResult = this.tokenList.filter((token) => {
-      return (
-        token.address &&
-        searchQuery &&
-        isEqualCaseInsensitive(token.address, searchQuery)
-      );
-    });
-    const results = [...addressSearchResult, ...fuseSearchResult];
-    this.props.onSearch({ searchQuery, results });
-  }
+  const handleSearch = useCallback(
+    (newSearchQuery) => {
+      setSearchQuery(newSearchQuery);
+      const fuseSearchResult = tokenSearchFuse.search(newSearchQuery);
+      const addressSearchResult = getTokens(tokenList).filter((token) => {
+        return (
+          token.address &&
+          searchQuery &&
+          isEqualCaseInsensitive(token.address, newSearchQuery)
+        );
+      });
+      const results = [...addressSearchResult, ...fuseSearchResult];
+      onSearch({ newSearchQuery, results });
+    },
+    [onSearch, searchQuery, tokenList, tokenSearchFuse],
+  );
 
-  clear() {
-    this.setState({ searchQuery: '' });
-  }
-
-  render() {
-    const { error } = this.props;
-    const { searchQuery } = this.state;
-    const { searchClassName } = this.props;
+  const clear = () => {
+    setSearchQuery('');
+  };
 
-    return (
-      <TextFieldSearch
-        className={searchClassName}
-        placeholder={this.context.t('searchTokens')}
-        value={searchQuery}
-        onChange={(e) => this.handleSearch(e.target.value)}
-        error={error}
-        autoFocus
-        autoComplete={false}
-        width={BlockSize.Full}
-        clearButtonOnClick={() => this.clear()}
-        clearButtonProps={{
-          size: Size.SM,
-        }}
-      />
-    );
-  }
+  return (
+    <TextFieldSearch
+      className={searchClassName}
+      placeholder={t('searchTokens')}
+      value={searchQuery}
+      onChange={(e) => handleSearch(e.target.value)}
```

Comment:
Using a wrapper anonymous function here kinda defeats the purpose of using a `useCallback` hook for this handler. `useCallback` prevents the function reference from changing each render, but that won't matter if we're wrapping it in an anonymous function that itself changes references each render.

Maybe better to drop the `useCallback` altogether, and leave this behaving as it did before. But if you want to avoid some additional renders, you'll want to drop this wrapper anonymous function, and the one used for `clear` (and introduce a `useCallback` call for `clear`).

Reply:
Dropping `useCallback` would be OK because this would be a very minor optimization, probably not impactful. And because this will rerender very often anyway no matter what we do, because `tokenList` is from background state, and the way state syncing works today results in every state update updating all object references, causing rerenders of anything referencing background state.

Reply:
I added the drop [here](https://github.com/MetaMask/metamask-extension/pull/22548/commits/83b4a7063250507068f252348bb54534117eeab3) and will do another MR to do improvement of this component

Reply:
changed [here](https://github.com/MetaMask/metamask-extension/pull/22548/commits/978bf0b881bd349eee7107c81cecc0068a636070)